[10.1.10] Outgoing request stream error on pool shutdown due to idle timeout

Using akka-http 10.1.10 on scala 2.13.0

I have a dedicated dispatcher configured for blocking calls when using a java API

```
auth-dispatcher {
  type = Dispatcher
  executor = "thread-pool-executor"
  thread-pool-executor {
    fixed-pool-size = 2
  }
  throughput = 1
}
```

Up until 10.1.9 this worked flawlessly. Since 10.1.10 it still seems to behave correctly but the log gets spammed with "Outgoing request stream error" messages everytime the idle-timeout kicks in
```
[INFO] [09/23/2019 12:32:07.431] [tsmAdmin-akka.actor.default-dispatcher-7] [tsmAdmin/Pool(shared->https://keycloak.example.com:443)] Pool shutting down because akka.http.host-connection-pool.idle-timeout triggered after 30 seconds.
[INFO] [09/23/2019 12:32:07.436] [tsmAdmin-akka.actor.default-dispatcher-7] [tsmAdmin/Pool(shared->https://keycloak.example.com:443)] Pool is now shutting down as requested.
[ERROR] [09/23/2019 12:32:07.451] [tsmAdmin-akka.actor.default-dispatcher-7] [akka.actor.ActorSystemImpl(tsmAdmin)] Outgoing request stream error (akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot$$anon$2: Pool slot was shut down)
[INFO] [09/23/2019 12:32:37.771] [tsmAdmin-akka.actor.default-dispatcher-5] [tsmAdmin/Pool(shared->https://keycloak.example.com:443)] Pool shutting down because akka.http.host-connection-pool.idle-timeout triggered after 30 seconds.
[INFO] [09/23/2019 12:32:37.771] [tsmAdmin-akka.actor.default-dispatcher-5] [tsmAdmin/Pool(shared->https://keycloak.example.com:443)] Pool is now shutting down as requested.
[ERROR] [09/23/2019 12:32:37.772] [tsmAdmin-akka.actor.default-dispatcher-5] [akka.actor.ActorSystemImpl(tsmAdmin)] Outgoing request stream error (akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot$$anon$2: Pool slot was shut down)
```

I can confirm this happening when doing the simplest things like `Http().singleRequest()`. The response entity is properly handled/discarded. 

The logs are full of:
```
[INFO] ..."Pool is now shutting down as requested." ...
[ERROR] ..."Pool slot was shut down" ... "Outgoing request stream error"...
```

I think the `idle-timeout timer` is being reinitialized after being previously canceled.

I added a minimal example here: https://github.com/markus1189/akka-http-10-1-10-2728

If you run this the output is:

```
[INFO] [09/24/2019 11:23:54.120] [default-akka.actor.default-dispatcher-2] [default/Pool(shared->http://akka.io:80)] Pool shutting down because akka.http.host-connection-pool.idle-timeout triggered after 30 seconds.
[INFO] [09/24/2019 11:23:54.121] [default-akka.actor.default-dispatcher-2] [default/Pool(shared->http://akka.io:80)] Pool is now shutting down as requested.
[ERROR] [09/24/2019 11:23:54.125] [default-akka.actor.default-dispatcher-2] [akka.actor.ActorSystemImpl(default)] Outgoing request stream error (akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot$$anon$2: Pool slot was shut down)
[INFO] [09/24/2019 11:24:28.839] [default-akka.actor.default-dispatcher-2] [default/Pool(shared->http://akka.io:80)] Pool shutting down because akka.http.host-connection-pool.idle-timeout triggered after 30 seconds.
[INFO] [09/24/2019 11:24:28.839] [default-akka.actor.default-dispatcher-2] [default/Pool(shared->http://akka.io:80)] Pool is now shutting down as requested.
[ERROR] [09/24/2019 11:24:28.840] [default-akka.actor.default-dispatcher-2] [akka.actor.ActorSystemImpl(default)] Outgoing request stream error (akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot$$anon$2: Pool slot was shut down)
... (and so on)
```

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolInterface.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolInterface.scala
@@ -130,7 +130,7 @@ private[http] object PoolInterface {
     }
 
     override protected def onTimer(timerKey: Any): Unit = {
-      log.info(s"Pool shutting down because akka.http.host-connection-pool.idle-timeout triggered after $idleTimeout.")
+      log.debug(s"Pool shutting down because akka.http.host-connection-pool.idle-timeout triggered after $idleTimeout.")
       requestShutdown(ShutdownReason.IdleTimeout)
     }
 
@@ -195,7 +195,7 @@ private[http] object PoolInterface {
     def shutdownIfRequestedAndPossible(): Unit =
       if (shuttingDown) {
         if (remainingRequested == 0) {
-          log.info("Pool is now shutting down as requested.")
+          log.debug("Pool is now shutting down as requested.")
           shutdownPromise.trySuccess(shuttingDownReason.get)
           completeStage()
         } else

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala
@@ -194,7 +194,13 @@ private[client] object NewHostConnectionPool {
           def isIdle: Boolean = state.isIdle
           def isConnected: Boolean = state.isConnected
           def shutdown(): Unit = {
-            closeConnection(Some(new IllegalStateException("Pool slot was shut down") with NoStackTrace))
+            // if the connection is idle, we just complete it regularly, otherwise, we forcibly tear it down
+            // with an error (which will be logged in OutgoingConnectionBlueprint, see `mapError` there).
+            val reason =
+              if (isIdle) None
+              else Some(new IllegalStateException("Pool slot was shut down") with NoStackTrace)
+
+            closeConnection(reason)
 
             state.onShutdown(this)
           }

--- a/akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala
+++ b/akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala
@@ -14,6 +14,8 @@ import scala.concurrent.duration._
 import scala.concurrent.{ Await, Future, Promise }
 import scala.util.{ Success, Try }
 import akka.actor.ActorSystem
+import akka.event.Logging
+import akka.event.Logging.LogEvent
 import akka.http.impl.engine.ws.ByteStringSinkProbe
 import akka.http.impl.util._
 import akka.http.scaladsl.Http.ServerBinding
@@ -472,6 +474,36 @@ class ClientServerSpec extends WordSpec with Matchers with BeforeAndAfterAll wit
         clientIn.expectComplete()
 
         binding.foreach(_.unbind())
+      }
+    }
+
+    "properly complete a simple request/response cycle using Http.singleRequest" in Utils.assertAllStagesStopped {
+      new TestSetup {
+        // make sure no log message above DEBUG are printed by that exchange
+        EventFilter.custom({ case l: LogEvent if l.level != Logging.DebugLevel => true }, 0).intercept {
+          val settings = ConnectionPoolSettings(system).withIdleTimeout(500.millis)
+
+          val request = HttpRequest(uri = s"http://$hostname:$port/abc")
+
+          val responseFut = Http().singleRequest(request, settings = settings)
+          val (serverIn, serverOut) = acceptConnection()
+
+          val serverInSub = serverIn.expectSubscription()
+          serverInSub.request(1)
+          serverIn.expectNext().uri shouldEqual Uri(s"http://$hostname:$port/abc")
+
+          val serverOutSub = serverOut.expectSubscription()
+          serverOutSub.expectRequest()
+          serverOutSub.sendNext(HttpResponse(entity = "yeah"))
+
+          val response = responseFut.awaitResult(1.second.dilated)
+          toStrict(response.entity) shouldEqual HttpEntity("yeah")
+
+          serverIn.expectComplete()
+          serverOutSub.expectCancellation()
+
+          binding.foreach(_.unbind())
+        }
       }
     }
 

